internal/speakerlist/speakerlist.go

```diff
@@ -80,36 +83,107 @@ func New(logger gokitlog.Logger, nodeName, bindAddr, bindPort, secret, namespace
 }
 
 // Start starts the needed goroutines
-func (sl *SpeakerList) Start(client *k8s.Client) error {
+func (sl *SpeakerList) Start(client *k8s.Client) {
 	sl.client = client
 
 	if sl.mList == nil {
-		return nil
+		return
 	}
 
 	go sl.watchEvents()
+	go sl.memberlistJoinLoop()
+}
 
-	iplist, err := sl.client.GetPodsIPs(sl.namespace, sl.labels)
-	if err != nil {
-		sl.l.Log("op", "startup", "error", err, "msg", "failed to get PodsIPs")
-		return err
+func (sl *SpeakerList) memberlistJoin() {
+	if sl.mList == nil {
+		return
+	}
+	joinList := sl.joinList()
+	if len(joinList) > 0 {
+		n, err := sl.mList.Join(joinList)
+		sl.l.Log("op", "ml-join", "msg", "Memberlist join", "nb joigned", n, "error ?", err, "join list", joinList)
+	}
+}
+
+func (sl *SpeakerList) joinList() []string {
+	joinList := []string{}
+	sl.Lock()
+	defer sl.Unlock()
+	if len(sl.k8sSpeakers) == 0 {
+		return joinList
+	}
+
+	mlSpeakers := map[string]bool{}
+	for _, n := range sl.mList.Members() {
+		mlSpeakers[n.Name] = true
+	}
+	for name, ip := range sl.k8sSpeakers {
+		if _, ok := mlSpeakers[name]; !ok {
+			joinList = append(joinList, ip)
+		}
+	}
+	return joinList
+}
+
+func (sl *SpeakerList) memberlistJoinLoop() {
+	ticker := time.NewTicker(10 * time.Second)
+	for {
+		select {
+		case <-sl.stopCh:
+			return
+		case <-ticker.C:
+			sl.memberlistJoin()
+		}
 	}
-	n, err := sl.mList.Join(iplist)
-	sl.l.Log("op", "startup", "msg", "Memberlist join", "nb joigned", n, "error ?", err)
+}
 
-	return nil
+// SetSpeakersK8S update the list of speakers from the K8S API point of view
+func (sl *SpeakerList) SetSpeakersK8S(eps *v1.Endpoints) {
+	newmap := map[string]string{}
+	for _, subset := range eps.Subsets {
+		for _, ep := range subset.Addresses {
+			newmap[*ep.NodeName] = ep.IP
+		}
+	}
+	sl.Lock()
+	oldmap := sl.k8sSpeakers
+	sl.Unlock()
+	if !reflect.DeepEqual(oldmap, newmap) {
+		sl.l.Log("op", "setSpeakersK8S", "msg", "New K8S Speakers list set", "old-list", oldmap, "new-list", newmap)
+		sl.Lock()
+		sl.k8sSpeakers = newmap
+		sl.Unlock()
+		sl.memberlistJoin()
+		// needed for split brain and disabled MemberList
+		sl.client.ForceSync()
+	}
 }
 
-// UsableSpeakers return a map of usable nodes
-func (sl *SpeakerList) UsableSpeakers() map[string]bool {
-	if sl.mList != nil {
-		activeNodes := map[string]bool{}
-		for _, n := range sl.mList.Members() {
-			activeNodes[n.Name] = true
+// UsableSpeakers return a map of usable nodes or an error
+func (sl *SpeakerList) UsableSpeakers() (map[string]bool, error) {
+	sl.Lock()
+	defer sl.Unlock()
+	if len(sl.k8sSpeakers) == 0 {
+		return nil, errors.New("k8sSpeakersNotSet")
+	}
+	activeNodes := map[string]bool{}
+	if sl.mList == nil {
+		for name := range sl.k8sSpeakers {
+			activeNodes[name] = true
 		}
-		return activeNodes
+		return activeNodes, nil
+	}
+
+	members := sl.mList.Members()
+	if len(sl.k8sSpeakers) > 1 && float64(len(members)) < float64(len(sl.k8sSpeakers))/2 {
```

Comment:
This means that if we lose half the workers running memberlist (let's say, N/2+1 workers are just down) **none** of the workers in the cluster will advertise any service as:
 * N/2+1 workers are just dead (let's say network cable unplugged)
 * This condition is false in the remaining workers that are up (as they are less than half)

And this will cause downtime for all service until we kubernetes returns less speakers in the query. But I'm not sure that will happen and if there is any way to recover from this, as I **think** kubernetes will show the speaker pods as running even when the node is down (maybe only when not using a CCM), because daemonset used to be tricky in this regard and have some gotchas. Have you tested if this is the case? (hopefully it is not and I just don't remember correctly).

Reply:
I'll test that, but I hope when the node is dead, the endpoint is not returned anymore

Reply:
When using Service / Endpoints
```
{"caller":"state.go:459","component":"MemberList","msg":"[INFO] memberlist: Suspect etienne-ks144 has failed, no acks received","ts":"2020-05-05T03:05:00.143308116Z"}
...
{"caller":"speakerlist.go:150","msg":"New K8S Speakers list set","new-list":{"etienne-ks141":"10.10.52.141","etienne-ks142":"10.10.52.142","etienne-ks143":"10.10.52.143"},"old-list":{"etienne-ks141":"10.10.52.141","etienne-ks142":"10.10.52.142","etienne-ks143":"10.10.52.143","etienne-ks144":"10.10.52.144"},"op":"setSpeakersK8S","ts":"2020-05-05T03:05:43.160324704Z"}
```

Now If I look at at the pods, it's running but not ready
```
  status:
    conditions:
    - lastProbeTime: null
      lastTransitionTime: "2020-05-05T03:02:39Z"
      status: "True"
      type: Initialized
    - lastProbeTime: null
      lastTransitionTime: "2020-05-05T03:05:43Z"
      status: "False"
      type: Ready
    - lastProbeTime: null
      lastTransitionTime: "2020-05-05T03:02:42Z"
      status: "True"
      type: ContainersReady
    - lastProbeTime: null
      lastTransitionTime: "2020-05-05T03:02:39Z"
      status: "True"
      type: PodScheduled
    containerStatuses:
    - containerID: containerd://88e8dd723e1d3afed97f679ea84f1e66d3386a984a071e7a8df513ff24a163d5
      image: docker.io/champtar/metalspeaker:c90537c5
      imageID: docker.io/champtar/metalspeaker@sha256:ea75717c67d0e80b25073105fb7150991d84356e7e7bc822b8bf3ce879aaa5b1
      lastState: {}
      name: speaker
      ready: true
      restartCount: 0
      state:
        running:
          startedAt: "2020-05-05T03:02:42Z"
    hostIP: 10.10.52.144
    phase: Running
    podIP: 10.10.52.144
    qosClass: BestEffort
    startTime: "2020-05-05T03:02:39Z"
```